File: backend/app/schemas/sonify.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class SonifyRegion(BaseModel):
    x: float = Field(ge=0, le=1)
    y: float = Field(ge=0, le=1)
    width: float = Field(gt=0, le=1)
    height: float = Field(gt=0, le=1)


class SonifyAnalysis(BaseModel):
    target_type: Literal["face", "faces", "object", "unknown"]
    confidence: float = Field(ge=0, le=1)
    region: SonifyRegion | None = None
    description: str = ""


class SonifyResponse(BaseModel):
    image_base64: str
    media_type: Literal["image/png", "image/jpeg", "image/webp"] = "image/png"
    width: int = Field(gt=0)
    height: int = Field(gt=0)
    analysis: SonifyAnalysis
    model: str

File: frontend/src/api/sonify.ts
import { API_BASE_URL } from './client'

export type SonifyResult = {
  image_base64: string
  media_type: string
  width: number
  height: number
  analysis: {
    target_type: 'face' | 'faces' | 'object' | 'unknown'
    confidence: number
    region: { x: number; y: number; width: number; height: number } | null
    description: string
  }
  model: string
}

export async function generateSonify(file: File, signal?: AbortSignal): Promise<SonifyResult> {
  const form = new FormData()
  form.append('target_image', file)
  const response = await fetch(`${API_BASE_URL}/sonify`, { method: 'POST', body: form, signal })
  if (!response.ok) {
    let message = `Generation failed with ${response.status}.`
    try {
      const body = await response.json() as { detail?: string }
      if (body.detail) message = body.detail
    } catch {
      // Keep the HTTP status when the server did not return JSON.
    }
    throw new Error(message)
  }
  return await response.json() as SonifyResult
}

export function sonifyImageUrl(result: SonifyResult): string {
  return `data:${result.media_type};base64,${result.image_base64}`
}
